# **AdvDS - Assignment 1: Binary Classification - NBA Player Prediction - Data Preparation**

Build a binary classification model that can predict if a rookie player will last at least 5 years in the NBA

**Student Name:** Nathan Fragar

**Student No. :** 93087548

**Team Name:** Group 1
* Nuwan Munasinghe
* Wenyingwuwy
* Nathan Fragar
* Sean Williams

## 1. Load and Discover Dataset

**[1.1]** Task: Import required packages: Pandas, Numpy, joblib

In [69]:
# Task: Import the pandas, numpy and joblib package
import pandas as pd
import numpy as np
import joblib as job

**[1.2]** Task: Identify correct directory

In [2]:
! ls -l ../data/raw

total 1912
-rwxrwxrwx 1 root root  41952 Feb  8 10:25  sample_submission.csv
-rwxrwxrwx 1 root root 322813 Feb  8 10:25  test.csv
-rwxrwxrwx 1 root root 445464 Feb 13 10:01  train.csv
-rwxrwxrwx 1 root root 695622 Feb  8 10:25 'train - orig - pre-clean - 20220213.csv'
-rwxrwxrwx 1 root root 445464 Feb 13 10:01 'train - post - clean - 20220213.csv'


**[1.3]** Task: Load Test and Train Files into Pandas Data Frames

In [3]:
# File URLs
file_path_train = "../data/raw/train.csv"
file_path_test= "../data/raw/test.csv"

# Load files into df_training and df_validation data frames
df_train = pd.read_csv(file_path_train)
df_test = pd.read_csv(file_path_test)

**[1.4]** Task: Display Train and Test Data Frames

In [4]:
# Task: Display first 5 rows of df_train
df_train.head()

Id  GP   MIN   PTS  FGM   FGA   FG%  3P Made  3PA   3P%  ...  FTA   FT%  \
0   4  83  18.8   8.3  3.1   6.6  47.0      0.1  0.4  25.0  ...  2.6  84.6   
1   8  56   6.4   1.6  0.6   1.7  35.3      0.2  0.5  40.0  ...  0.4  50.0   
2   9  75  30.5  13.6  5.0  11.4  43.9      0.3  1.1  27.3  ...  3.9  71.8   
3  10  71  15.9   6.4  2.4   5.5  43.6      0.2  0.4  50.0  ...  1.6  81.3   
4  16  40  17.1   6.7  2.3   5.8  39.7      0.3  0.9  33.3  ...  2.0  70.0   

   OREB  DREB  REB  AST  STL  BLK  TOV  TARGET_5Yrs  
0   0.8   2.3  3.0  0.8  0.7  0.2  1.0            1  
1   0.6   0.5  0.9  0.2  0.1  0.3  0.3            1  
2   0.9   4.3  5.3  2.2  0.6  0.7  1.9            1  
3   0.9   2.7  3.7  1.3  0.6  0.6  1.4            1  
4   1.0   2.6  3.3  0.4  0.5  0.4  1.2            1  

[5 rows x 21 columns]

In [5]:
# Task: Display first 5 rows of df_train
df_test.head()

Id  GP   MIN   PTS  FGM   FGA   FG%  3P Made  3PA   3P%  FTM  FTA   FT%  \
0     1  56   9.1   4.0  1.6   3.7  43.7      0.1  0.3   7.3  0.7  1.2  63.4   
1  8194  43  19.3  10.1  3.7   8.1  46.0      0.6  1.7  35.1  1.8  2.5  75.3   
2     3  82  33.9  11.3  4.9  10.6  45.6      0.5  1.9  44.8  1.8  2.7  71.2   
3  8196  86  44.7  18.8  6.8  15.9  42.9      0.5  1.8  13.5  4.5  6.3  70.9   
4  8197  58  12.3   4.7  1.6   4.0  40.0      0.5  1.7  38.7  1.1  1.3  76.9   

   OREB  DREB  REB  AST  STL  BLK  TOV  
0   1.2   0.8  1.7  0.4  0.2  0.3  0.8  
1   0.5   0.9  1.5  3.5  0.6 -0.0  1.8  
2   1.3   3.3  4.5  2.5  1.3  0.3  2.0  
3   1.5   3.2  5.0  4.1  0.9  0.1  3.6  
4   0.2   0.6  0.9  1.5  0.5 -0.4  0.9

**Observation:** Both datasets are identical with the exception the Test Dataset does not have the Target Variable. This is to be expected

**[1.5]** Task: Display Train and Test Shape

In [83]:
# Task: Display shape of df_train
df_train.shape

(5256, 21)

In [84]:
# Task: Display shape of df_test
df_test.shape

(3799, 20)

**[1.6]** Task: Display descriptive statistics for Train and Test Datasets

In [85]:
# Task: Describe df_train
df_train.describe()

Id           GP          MIN          PTS          FGM  \
count   5256.000000  5256.000000  5256.000000  5256.000000  5256.000000   
mean    6923.773973    65.010274    20.794844     8.238451     3.160940   
std     3981.202752    16.760185     9.115862     4.481536     1.760233   
min        4.000000     3.000000     3.800000     0.900000     0.300000   
25%     3474.500000    54.000000    14.100000     4.900000     1.800000   
50%     6904.000000    65.000000    19.200000     7.400000     2.800000   
75%    10364.750000    76.000000    25.900000    10.600000     4.100000   
max    13798.000000   123.000000    73.800000    34.200000    13.100000   

               FGA          FG%      3P Made          3PA          3P%  ...  \
count  5256.000000  5256.000000  5256.000000  5256.000000  5256.000000  ...   
mean      7.042295    44.478196     0.404072     1.202055    32.443588  ...   
std       3.721512     5.681455     0.284149     0.783867    15.316130  ...   
min       0.900000    25.000000     0.000000     0.000000     0.000000  ...   
25%       4.300000    40.500000     0.200000     0.600000    27.800000  ...   
50%       6.300000    44.400000     0.400000     1.100000    33.300000  ...   
75%       9.000000    48.200000     0.600000     1.700000    37.500000  ...   
max      28.900000    66.700000     1.700000     4.700000   100.000000  ...   

               FTA          FT%         OREB         DREB          REB  \
count  5256.000000  5256.000000  5256.000000  5256.000000  5256.000000   
mean      2.130613    71.678044     1.122127     2.419749     3.550666   
std       1.283860    10.785198     0.735496     1.397683     2.050265   
min       0.000000     0.000000     0.000000     0.200000     0.400000   
25%       1.200000    65.200000     0.600000     1.400000     2.100000   
50%       1.900000    71.400000     1.000000     2.100000     3.100000   
75%       2.800000    77.800000     1.500000     3.100000     4.600000   
max      11.100000   100.000000     5.500000    10.800000    14.600000   

               AST          STL          BLK          TOV  TARGET_5Yrs  
count  5256.000000  5256.000000  5256.000000  5256.000000  5256.000000  
mean      1.785255     0.717751     0.339783     1.373516     0.848364  
std       1.389877     0.422325     0.579887     0.750149     0.358702  
min       0.000000     0.000000     0.000000     0.100000     0.000000  
25%       0.800000     0.400000     0.100000     0.800000     1.000000  
50%       1.400000     0.600000     0.200000     1.200000     1.000000  
75%       2.400000     0.900000     0.400000     1.800000     1.000000  
max      12.800000     3.600000    18.900000     5.300000     1.000000  

[8 rows x 21 columns]

**Observations**

**Outlier - Minimum Values** - Negative for GP, 3P Made, 3PA, FTM, BLK
* GP: -8 games
* 3P Made: -1.1 
* 3PA: -3.1
* FTM: 
* BLK: -17.9

**Outlier - Maximum Values**
* MIN - A max value of 73.8 minutes played per game appears high, particularly with 75% at 23.5 - may have some outlier records
* a NBA Game has 48 minutes per game


In [6]:
# Task: Describe df_train
df_test.describe()

Id           GP          MIN          PTS          FGM  \
count   3799.000000  3799.000000  3799.000000  3799.000000  3799.000000   
mean    7010.614109    62.853909    18.650224     7.328034     2.835404   
std     3954.173641    17.151740     8.727259     4.294724     1.688427   
min        1.000000     6.000000     3.700000     0.700000     0.300000   
25%     3644.000000    51.000000    12.200000     4.200000     1.600000   
50%     7062.000000    63.000000    17.000000     6.400000     2.500000   
75%    10402.500000    74.000000    23.300000     9.400000     3.700000   
max    13792.000000   126.000000    68.000000    33.000000    13.400000   

               FGA          FG%      3P Made          3PA          3P%  \
count  3799.000000  3799.000000  3799.000000  3799.000000  3799.000000   
mean      6.302580    44.599079     0.255962     0.796920    19.234746   
std       3.579221     6.040168     0.380987     1.052862    15.968989   
min       0.800000    25.100000    -1.000000    -2.700000   -38.000000   
25%       3.700000    40.500000     0.000000     0.100000     8.500000   
50%       5.500000    44.600000     0.300000     0.800000    19.400000   
75%       8.100000    48.500000     0.500000     1.500000    30.250000   
max      26.200000    74.600000     1.600000     4.300000    73.800000   

               FTM          FTA          FT%         OREB         DREB  \
count  3799.000000  3799.000000  3799.000000  3799.000000  3799.000000   
mean      1.399842     1.953567    71.612924     1.096025     2.179495   
std       0.926140     1.250376    10.457336     0.785678     1.371935   
min       0.000000     0.000000    23.700000     0.000000     0.200000   
25%       0.700000     1.000000    65.000000     0.500000     1.200000   
50%       1.200000     1.700000    71.500000     0.900000     1.900000   
75%       1.900000     2.600000    78.000000     1.500000     2.900000   
max       7.800000     9.800000   127.100000     6.900000    12.000000   

               REB          AST          STL          BLK          TOV  
count  3799.000000  3799.000000  3799.000000  3799.000000  3799.000000  
mean      3.275783     1.636483     0.653593     0.257726     1.257910  
std       2.070646     1.335496     0.410573     0.639660     0.712449  
min       0.300000     0.000000     0.000000    -7.100000     0.100000  
25%       1.800000     0.600000     0.400000     0.100000     0.700000  
50%       2.800000     1.300000     0.600000     0.200000     1.100000  
75%       4.300000     2.300000     0.900000     0.400000     1.600000  
max      18.500000     9.000000     2.700000    14.800000     5.200000

**Obervations**
* No Negative minimums
* Outlier Maximum Values - MIN - a max value of 68 mins per game appears high
* Mean and Standard Deviation between train and test appear to be reasonably well aligned


**[1.7]** Task: Identify Missing Values

In [7]:
# Task: Display Summary (Info)
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5256 entries, 0 to 5255
Data columns (total 21 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Id           5256 non-null   int64  
 1   GP           5256 non-null   int64  
 2   MIN          5256 non-null   float64
 3   PTS          5256 non-null   float64
 4   FGM          5256 non-null   float64
 5   FGA          5256 non-null   float64
 6   FG%          5256 non-null   float64
 7   3P Made      5256 non-null   float64
 8   3PA          5256 non-null   float64
 9   3P%          5256 non-null   float64
 10  FTM          5256 non-null   float64
 11  FTA          5256 non-null   float64
 12  FT%          5256 non-null   float64
 13  OREB         5256 non-null   float64
 14  DREB         5256 non-null   float64
 15  REB          5256 non-null   float64
 16  AST          5256 non-null   float64
 17  STL          5256 non-null   float64
 18  BLK          5256 non-null   float64
 19  TOV   

**Oberservation** 
There are no missing values in this dataset

**[1.8]** Task: Investigate Column Details - Find Min Values for each columns

In [88]:
df_train.min()

Id              4.0
GP              3.0
MIN             3.8
PTS             0.9
FGM             0.3
FGA             0.9
FG%            25.0
3P Made         0.0
3PA             0.0
3P%             0.0
FTM             0.0
FTA             0.0
FT%             0.0
OREB            0.0
DREB            0.2
REB             0.4
AST             0.0
STL             0.0
BLK             0.0
TOV             0.1
TARGET_5Yrs     0.0
dtype: float64

**[1.8.1]** Task: Find Rows where GP (Games Played) is negative

In [8]:
df_train_mask_GP = df_train['GP'] < 0
df_train_GP_less_zero = df_train[df_train_mask_GP]
print(df_train_GP_less_zero)

Empty DataFrame
Columns: [Id, GP, MIN, PTS, FGM, FGA, FG%, 3P Made, 3PA, 3P%, FTM, FTA, FT%, OREB, DREB, REB, AST, STL, BLK, TOV, TARGET_5Yrs]
Index: []

[0 rows x 21 columns]


**[1.8.2]** Task: Find Rows where 3P Made (3 Points Made) is negative

In [90]:
# Identify how many rows are positive, negative and zero
#df_train_mask_3PMade = df_train['3P Made'] > 0
#df_train_mask_3PMade = df_train['3P Made'] == 0
df_train_mask_3PMade = df_train['3P Made'] < 0
df_train_3PMade_less_zero = df_train[df_train_mask_3PMade]
print(df_train_3PMade_less_zero[['Id','GP','MIN','3P Made','3PA','3P%']])

Empty DataFrame
Columns: [Id, GP, MIN, 3P Made, 3PA, 3P%]
Index: []


**Observed Statistics - Negative Values for 3P Made**
* Negative - 1629 (20.4%)
* Zero - 673 (8.4%)
* Positive - 5698 (71.2%)


**[1.8.3]** Task: Find Rows where 3PA (3 Points Attempted) is negative

In [91]:
# Identify how many rows are positive, negative and zero
#df_train_mask_3PMade = df_train['3PA'] > 0
#df_train_mask_3PMade = df_train['3PA'] == 0
df_train_mask_3PMade = (df_train['3PA'] > 0)
df_train_3PMade_less_zero = df_train[df_train_mask_3PMade]
print(df_train_3PMade_less_zero[['Id','GP','MIN','3P Made','3PA','3P%']])

         Id   GP   MIN  3P Made  3PA   3P%
0         4   83  18.8      0.1  0.4  25.0
1         8   56   6.4      0.2  0.5  40.0
2         9   75  30.5      0.3  1.1  27.3
3        10   71  15.9      0.2  0.4  50.0
4        16   40  17.1      0.3  0.9  33.3
...     ...  ...   ...      ...  ...   ...
5251  13794   49  15.4      0.0  0.2   0.0
5252  13795  108  33.0      0.2  0.5  40.0
5253  13796   48  17.6      0.5  1.8  27.8
5254  13797   45  14.4      0.3  0.8  37.5
5255  13798   62  15.8      0.3  0.8  37.5

[5155 rows x 6 columns]


**Observated Statistics**
* Negative - 1658 (20.7%)
* Zero - 220 (2.8%)
* Positive - 6342  (76.5%)

**Observation:** The 3 point statistics do not look reliable.

**Proposal:** Remove 3 point statistics from initial analysis

**Notes**
The following columns do not appear to have reliable information, they do not fit the profile of the statistic
* 3P Made - Negative Numbers should not be possible
* 3PA - Points Attempted should always be the same or greater than Points Made
* 3P% - Calculation of % should be (Points Made / Points Attempted) within a certain tolerance of error (due to rounding)
* FTM - Negative Numbers should not be possible
* FTA - Points Attempted should always be the same or greater than Points Made
* FT% 
* BLK  - Negative Numbers should not be possible


**[1.9]** Task: Check for duplicates

In [92]:
# Identify Rows that are Duplicated
dup = df_train.duplicated()
df_train[dup]

Empty DataFrame
Columns: [Id, GP, MIN, PTS, FGM, FGA, FG%, 3P Made, 3PA, 3P%, FTM, FTA, FT%, OREB, DREB, REB, AST, STL, BLK, TOV, TARGET_5Yrs]
Index: []

[0 rows x 21 columns]

In [93]:
# Identify Rows that are Duplicated
dup = df_test.duplicated()
df_test[dup]

Empty DataFrame
Columns: [Id, GP, MIN, PTS, FGM, FGA, FG%, 3P Made, 3PA, 3P%, FTM, FTA, FT%, OREB, DREB, REB, AST, STL, BLK, TOV]
Index: []

## 2. Feature Engineering

No creation of new features at this time.

**[2.1]** Create copy of Training and Test Datasets


In [9]:
# Create copy of df_train and df_test
df_train_cleaned = df_train
df_test_cleaned = df_test

**[2.2]** Task:Create Training and Test Datasets

Drop Columns: 3P Made,3PA ,3P%, FTM,FTA,FT%, ,BLK

In [95]:
# Drop Columns for Train and Test Datasets
#df_train_cleaned = df_train_cleaned.drop(['3P Made','3PA','3P%','FTM','FTA','FT%','BLK'],axis=1)
#df_test_cleaned = df_test_cleaned.drop(['3P Made','3PA','3P%','FTM','FTA','FT%','BLK'],axis=1)

Task: Display df_train_cleaned Information

In [10]:
# Display df_train_cleaned information
df_train_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5256 entries, 0 to 5255
Data columns (total 21 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Id           5256 non-null   int64  
 1   GP           5256 non-null   int64  
 2   MIN          5256 non-null   float64
 3   PTS          5256 non-null   float64
 4   FGM          5256 non-null   float64
 5   FGA          5256 non-null   float64
 6   FG%          5256 non-null   float64
 7   3P Made      5256 non-null   float64
 8   3PA          5256 non-null   float64
 9   3P%          5256 non-null   float64
 10  FTM          5256 non-null   float64
 11  FTA          5256 non-null   float64
 12  FT%          5256 non-null   float64
 13  OREB         5256 non-null   float64
 14  DREB         5256 non-null   float64
 15  REB          5256 non-null   float64
 16  AST          5256 non-null   float64
 17  STL          5256 non-null   float64
 18  BLK          5256 non-null   float64
 19  TOV   

In [11]:
df_test_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3799 entries, 0 to 3798
Data columns (total 20 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Id       3799 non-null   int64  
 1   GP       3799 non-null   int64  
 2   MIN      3799 non-null   float64
 3   PTS      3799 non-null   float64
 4   FGM      3799 non-null   float64
 5   FGA      3799 non-null   float64
 6   FG%      3799 non-null   float64
 7   3P Made  3799 non-null   float64
 8   3PA      3799 non-null   float64
 9   3P%      3799 non-null   float64
 10  FTM      3799 non-null   float64
 11  FTA      3799 non-null   float64
 12  FT%      3799 non-null   float64
 13  OREB     3799 non-null   float64
 14  DREB     3799 non-null   float64
 15  REB      3799 non-null   float64
 16  AST      3799 non-null   float64
 17  STL      3799 non-null   float64
 18  BLK      3799 non-null   float64
 19  TOV      3799 non-null   float64
dtypes: float64(18), int64(2)
memory usage: 593.7 KB


**[2.3]** Task: Import and Normalise all statistic columns

In [12]:
# Import MinMaxScaler
from sklearn.preprocessing import MinMaxScaler
# Instantiate Scaler
scaler = MinMaxScaler()

In [13]:
# Create df_train_cleaned_scaled_cols with columns to be scaled 
df_train_cleaned_scaled_cols = df_train_cleaned[['GP','MIN','PTS','FGM','FGA','FG%','OREB','DREB','REB','AST','STL','TOV','3P Made','3PA','3P%','FTM','FTA','FT%','BLK']]
df_test_cleaned_scaled_cols = df_test_cleaned[['GP','MIN','PTS','FGM','FGA','FG%','OREB','DREB','REB','AST','STL','TOV','3P Made','3PA','3P%','FTM','FTA','FT%','BLK']]
df_train_cleaned_scaled_cols.head()

GP   MIN   PTS  FGM   FGA   FG%  OREB  DREB  REB  AST  STL  TOV  3P Made  \
0  83  18.8   8.3  3.1   6.6  47.0   0.8   2.3  3.0  0.8  0.7  1.0      0.1   
1  56   6.4   1.6  0.6   1.7  35.3   0.6   0.5  0.9  0.2  0.1  0.3      0.2   
2  75  30.5  13.6  5.0  11.4  43.9   0.9   4.3  5.3  2.2  0.6  1.9      0.3   
3  71  15.9   6.4  2.4   5.5  43.6   0.9   2.7  3.7  1.3  0.6  1.4      0.2   
4  40  17.1   6.7  2.3   5.8  39.7   1.0   2.6  3.3  0.4  0.5  1.2      0.3   

   3PA   3P%  FTM  FTA   FT%  BLK  
0  0.4  25.0  2.2  2.6  84.6  0.2  
1  0.5  40.0  0.2  0.4  50.0  0.3  
2  1.1  27.3  2.8  3.9  71.8  0.7  
3  0.4  50.0  1.3  1.6  81.3  0.6  
4  0.9  33.3  1.4  2.0  70.0  0.4

In [14]:
# Scale Columns
df_train_cleaned_scaled = scaler.fit_transform(df_train_cleaned_scaled_cols)
df_test_cleaned_scaled = scaler.fit_transform(df_test_cleaned_scaled_cols)

In [15]:
# Convert numpy arrany df_train_cleaned_scaled values back to df_train_cleaned_scaled data frame 
df_train_cleaned_scaled = pd.DataFrame(np.squeeze(df_train_cleaned_scaled), columns=df_train_cleaned_scaled_cols.columns) 
df_test_cleaned_scaled = pd.DataFrame(np.squeeze(df_test_cleaned_scaled), columns=df_test_cleaned_scaled_cols.columns) 

In [16]:
df_train_cleaned_scaled.describe()

GP          MIN          PTS          FGM          FGA  \
count  5256.000000  5256.000000  5256.000000  5256.000000  5256.000000   
mean      0.516752     0.242783     0.220374     0.223511     0.219368   
std       0.139668     0.130227     0.134581     0.137518     0.132911   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.425000     0.147143     0.120120     0.117188     0.121429   
50%       0.516667     0.220000     0.195195     0.195312     0.192857   
75%       0.608333     0.315714     0.291291     0.296875     0.289286   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

               FG%         OREB         DREB          REB          AST  \
count  5256.000000  5256.000000  5256.000000  5256.000000  5256.000000   
mean      0.467103     0.204023     0.209410     0.221878     0.139473   
std       0.136246     0.133727     0.131857     0.144385     0.108584   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.371703     0.109091     0.113208     0.119718     0.062500   
50%       0.465228     0.181818     0.179245     0.190141     0.109375   
75%       0.556355     0.272727     0.273585     0.295775     0.187500   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

               STL          TOV      3P Made          3PA          3P%  \
count  5256.000000  5256.000000  5256.000000  5256.000000  5256.000000   
mean      0.199375     0.244907     0.237689     0.255756     0.324436   
std       0.117313     0.144259     0.167147     0.166780     0.153161   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.111111     0.134615     0.117647     0.127660     0.278000   
50%       0.166667     0.211538     0.235294     0.234043     0.333000   
75%       0.250000     0.326923     0.352941     0.361702     0.375000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

               FTM          FTA          FT%          BLK  
count  5256.000000  5256.000000  5256.000000  5256.000000  
mean      0.189683     0.191947     0.716780     0.017978  
std       0.118475     0.115663     0.107852     0.030682  
min       0.000000     0.000000     0.000000     0.000000  
25%       0.098765     0.108108     0.652000     0.005291  
50%       0.172840     0.171171     0.714000     0.010582  
75%       0.246914     0.252252     0.778000     0.021164  
max       1.000000     1.000000     1.000000     1.000000

In [17]:
df_train_cleaned.describe()

Id           GP          MIN          PTS          FGM  \
count   5256.000000  5256.000000  5256.000000  5256.000000  5256.000000   
mean    6923.773973    65.010274    20.794844     8.238451     3.160940   
std     3981.202752    16.760185     9.115862     4.481536     1.760233   
min        4.000000     3.000000     3.800000     0.900000     0.300000   
25%     3474.500000    54.000000    14.100000     4.900000     1.800000   
50%     6904.000000    65.000000    19.200000     7.400000     2.800000   
75%    10364.750000    76.000000    25.900000    10.600000     4.100000   
max    13798.000000   123.000000    73.800000    34.200000    13.100000   

               FGA          FG%      3P Made          3PA          3P%  ...  \
count  5256.000000  5256.000000  5256.000000  5256.000000  5256.000000  ...   
mean      7.042295    44.478196     0.404072     1.202055    32.443588  ...   
std       3.721512     5.681455     0.284149     0.783867    15.316130  ...   
min       0.900000    25.000000     0.000000     0.000000     0.000000  ...   
25%       4.300000    40.500000     0.200000     0.600000    27.800000  ...   
50%       6.300000    44.400000     0.400000     1.100000    33.300000  ...   
75%       9.000000    48.200000     0.600000     1.700000    37.500000  ...   
max      28.900000    66.700000     1.700000     4.700000   100.000000  ...   

               FTA          FT%         OREB         DREB          REB  \
count  5256.000000  5256.000000  5256.000000  5256.000000  5256.000000   
mean      2.130613    71.678044     1.122127     2.419749     3.550666   
std       1.283860    10.785198     0.735496     1.397683     2.050265   
min       0.000000     0.000000     0.000000     0.200000     0.400000   
25%       1.200000    65.200000     0.600000     1.400000     2.100000   
50%       1.900000    71.400000     1.000000     2.100000     3.100000   
75%       2.800000    77.800000     1.500000     3.100000     4.600000   
max      11.100000   100.000000     5.500000    10.800000    14.600000   

               AST          STL          BLK          TOV  TARGET_5Yrs  
count  5256.000000  5256.000000  5256.000000  5256.000000  5256.000000  
mean      1.785255     0.717751     0.339783     1.373516     0.848364  
std       1.389877     0.422325     0.579887     0.750149     0.358702  
min       0.000000     0.000000     0.000000     0.100000     0.000000  
25%       0.800000     0.400000     0.100000     0.800000     1.000000  
50%       1.400000     0.600000     0.200000     1.200000     1.000000  
75%       2.400000     0.900000     0.400000     1.800000     1.000000  
max      12.800000     3.600000    18.900000     5.300000     1.000000  

[8 rows x 21 columns]

In [18]:
df_train_cleaned = df_train_cleaned.drop(['GP','MIN','PTS','FGM','FGA','FG%','OREB','DREB','REB','AST','STL','TOV','3P Made','3PA','3P%','FTM','FTA','FT%','BLK'],axis=1)
df_test_cleaned = df_test_cleaned.drop(['GP','MIN','PTS','FGM','FGA','FG%','OREB','DREB','REB','AST','STL','TOV','3P Made','3PA','3P%','FTM','FTA','FT%','BLK'],axis=1)

In [19]:
df_train_cleaned = pd.concat([df_train_cleaned, df_train_cleaned_scaled], axis=1)
df_test_cleaned = pd.concat([df_test_cleaned, df_test_cleaned_scaled], axis=1)

In [20]:
df_train_cleaned.head(10)

Id  TARGET_5Yrs        GP       MIN       PTS       FGM       FGA  \
0   4            1  0.666667  0.214286  0.222222  0.218750  0.203571   
1   8            1  0.441667  0.037143  0.021021  0.023438  0.028571   
2   9            1  0.600000  0.381429  0.381381  0.367188  0.375000   
3  10            1  0.566667  0.172857  0.165165  0.164062  0.164286   
4  16            1  0.308333  0.190000  0.174174  0.156250  0.175000   
5  19            1  0.291667  0.292857  0.333333  0.343750  0.317857   
6  21            1  0.575000  0.174286  0.168168  0.171875  0.157143   
7  22            1  0.700000  0.341429  0.441441  0.367188  0.457143   
8  24            0  0.583333  0.177143  0.132132  0.148438  0.189286   
9  26            1  0.591667  0.142857  0.126126  0.093750  0.085714   

        FG%      OREB      DREB  ...       AST       STL       TOV   3P Made  \
0  0.527578  0.145455  0.198113  ...  0.062500  0.194444  0.173077  0.058824   
1  0.247002  0.109091  0.028302  ...  0.015625  0.027778  0.038462  0.117647   
2  0.453237  0.163636  0.386792  ...  0.171875  0.166667  0.346154  0.176471   
3  0.446043  0.163636  0.235849  ...  0.101562  0.166667  0.250000  0.117647   
4  0.352518  0.181818  0.226415  ...  0.031250  0.138889  0.211538  0.176471   
5  0.551559  0.254545  0.235849  ...  0.070312  0.250000  0.230769  0.352941   
6  0.532374  0.109091  0.169811  ...  0.070312  0.166667  0.153846  0.294118   
7  0.275779  0.072727  0.141509  ...  0.234375  0.472222  0.365385  0.705882   
8  0.251799  0.090909  0.207547  ...  0.156250  0.138889  0.134615  0.352941   
9  0.491607  0.109091  0.320755  ...  0.046875  0.111111  0.192308  0.294118   

        3PA    3P%       FTM       FTA    FT%       BLK  
0  0.085106  0.250  0.271605  0.234234  0.846  0.010582  
1  0.106383  0.400  0.024691  0.036036  0.500  0.015873  
2  0.234043  0.273  0.345679  0.351351  0.718  0.037037  
3  0.085106  0.500  0.160494  0.144144  0.813  0.031746  
4  0.191489  0.333  0.172840  0.180180  0.700  0.021164  
5  0.361702  0.353  0.283951  0.279279  0.742  0.031746  
6  0.297872  0.357  0.148148  0.126126  0.857  0.021164  
7  0.765957  0.333  0.407407  0.297297  1.000  0.000000  
8  0.404255  0.316  0.074074  0.072072  0.750  0.005291  
9  0.340426  0.313  0.259259  0.198198  0.955  0.010582  

[10 rows x 21 columns]

In [21]:
df_test_cleaned.head(10)

Id        GP       MIN       PTS       FGM       FGA       FG%      OREB  \
0     1  0.416667  0.083981  0.102167  0.099237  0.114173  0.375758  0.173913   
1  8194  0.308333  0.242613  0.291022  0.259542  0.287402  0.422222  0.072464   
2     3  0.633333  0.469673  0.328173  0.351145  0.385827  0.414141  0.188406   
3  8196  0.666667  0.637636  0.560372  0.496183  0.594488  0.359596  0.217391   
4  8197  0.433333  0.133748  0.123839  0.099237  0.125984  0.301010  0.028986   
5     6  0.441667  0.192846  0.198142  0.190840  0.145669  0.745455  0.260870   
6  8198  0.325000  0.071540  0.077399  0.083969  0.078740  0.430303  0.072464   
7  8202  0.391667  0.130638  0.123839  0.114504  0.129921  0.383838  0.086957   
8    14  0.291667  0.278383  0.222910  0.213740  0.204724  0.494949  0.347826   
9  8207  0.158333  0.097978  0.058824  0.053435  0.059055  0.402020  0.057971   

       DREB       REB       AST       STL       TOV   3P Made       3PA  \
0  0.050847  0.076923  0.044444  0.074074  0.137255  0.423077  0.428571   
1  0.059322  0.065934  0.388889  0.222222  0.333333  0.615385  0.628571   
2  0.262712  0.230769  0.277778  0.481481  0.372549  0.576923  0.657143   
3  0.254237  0.258242  0.455556  0.333333  0.686275  0.576923  0.642857   
4  0.033898  0.032967  0.166667  0.185185  0.156863  0.576923  0.628571   
5  0.203390  0.236264  0.066667  0.222222  0.156863  0.346154  0.300000   
6  0.025424  0.049451  0.022222  0.037037  0.058824  0.192308  0.157143   
7  0.076271  0.071429  0.088889  0.148148  0.098039  0.423077  0.414286   
8  0.203390  0.263736  0.155556  0.185185  0.254902  0.500000  0.514286   
9  0.110169  0.082418  0.044444  0.074074  0.078431  0.461538  0.485714   

        3P%       FTM       FTA       FT%       BLK  
0  0.405188  0.089744  0.122449  0.383946  0.337900  
1  0.653846  0.230769  0.255102  0.499033  0.324201  
2  0.740608  0.230769  0.275510  0.459381  0.337900  
3  0.460644  0.576923  0.642857  0.456480  0.328767  
4  0.686047  0.141026  0.132653  0.514507  0.305936  
5  0.370304  0.217949  0.275510  0.337524  0.342466  
6  0.419499  0.064103  0.071429  0.441006  0.333333  
7  0.569767  0.115385  0.102041  0.546422  0.315068  
8  0.207513  0.230769  0.224490  0.521277  0.342466  
9  0.444544  0.051282  0.051020  0.584139  0.333333

In [22]:
df_train_cleaned.shape

(5256, 21)

In [23]:
df_test_cleaned.shape

(3799, 20)

In [24]:
df_train_cleaned.describe()

Id  TARGET_5Yrs           GP          MIN          PTS  \
count   5256.000000  5256.000000  5256.000000  5256.000000  5256.000000   
mean    6923.773973     0.848364     0.516752     0.242783     0.220374   
std     3981.202752     0.358702     0.139668     0.130227     0.134581   
min        4.000000     0.000000     0.000000     0.000000     0.000000   
25%     3474.500000     1.000000     0.425000     0.147143     0.120120   
50%     6904.000000     1.000000     0.516667     0.220000     0.195195   
75%    10364.750000     1.000000     0.608333     0.315714     0.291291   
max    13798.000000     1.000000     1.000000     1.000000     1.000000   

               FGM          FGA          FG%         OREB         DREB  ...  \
count  5256.000000  5256.000000  5256.000000  5256.000000  5256.000000  ...   
mean      0.223511     0.219368     0.467103     0.204023     0.209410  ...   
std       0.137518     0.132911     0.136246     0.133727     0.131857  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%       0.117188     0.121429     0.371703     0.109091     0.113208  ...   
50%       0.195312     0.192857     0.465228     0.181818     0.179245  ...   
75%       0.296875     0.289286     0.556355     0.272727     0.273585  ...   
max       1.000000     1.000000     1.000000     1.000000     1.000000  ...   

               AST          STL          TOV      3P Made          3PA  \
count  5256.000000  5256.000000  5256.000000  5256.000000  5256.000000   
mean      0.139473     0.199375     0.244907     0.237689     0.255756   
std       0.108584     0.117313     0.144259     0.167147     0.166780   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.062500     0.111111     0.134615     0.117647     0.127660   
50%       0.109375     0.166667     0.211538     0.235294     0.234043   
75%       0.187500     0.250000     0.326923     0.352941     0.361702   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

               3P%          FTM          FTA          FT%          BLK  
count  5256.000000  5256.000000  5256.000000  5256.000000  5256.000000  
mean      0.324436     0.189683     0.191947     0.716780     0.017978  
std       0.153161     0.118475     0.115663     0.107852     0.030682  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%       0.278000     0.098765     0.108108     0.652000     0.005291  
50%       0.333000     0.172840     0.171171     0.714000     0.010582  
75%       0.375000     0.246914     0.252252     0.778000     0.021164  
max       1.000000     1.000000     1.000000     1.000000     1.000000  

[8 rows x 21 columns]

In [25]:
df_test_cleaned.describe()

Id           GP          MIN          PTS          FGM  \
count   3799.000000  3799.000000  3799.000000  3799.000000  3799.000000   
mean    7010.614109     0.473783     0.232507     0.205202     0.193542   
std     3954.173641     0.142931     0.135727     0.132964     0.128888   
min        1.000000     0.000000     0.000000     0.000000     0.000000   
25%     3644.000000     0.375000     0.132193     0.108359     0.099237   
50%     7062.000000     0.475000     0.206843     0.176471     0.167939   
75%    10402.500000     0.566667     0.304821     0.269350     0.259542   
max    13792.000000     1.000000     1.000000     1.000000     1.000000   

               FGA          FG%         OREB         DREB          REB  \
count  3799.000000  3799.000000  3799.000000  3799.000000  3799.000000   
mean      0.216637     0.393921     0.158844     0.167754     0.163505   
std       0.140914     0.122024     0.113866     0.116266     0.113772   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.114173     0.311111     0.072464     0.084746     0.082418   
50%       0.185039     0.393939     0.130435     0.144068     0.137363   
75%       0.287402     0.472727     0.217391     0.228814     0.219780   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

               AST          STL          TOV      3P Made          3PA  \
count  3799.000000  3799.000000  3799.000000  3799.000000  3799.000000   
mean      0.181831     0.242072     0.227041     0.483062     0.499560   
std       0.148388     0.152064     0.139696     0.146533     0.150409   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.066667     0.148148     0.117647     0.384615     0.400000   
50%       0.144444     0.222222     0.196078     0.500000     0.500000   
75%       0.255556     0.333333     0.294118     0.576923     0.600000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

               3P%          FTM          FTA          FT%          BLK  
count  3799.000000  3799.000000  3799.000000  3799.000000  3799.000000  
mean      0.511939     0.179467     0.199344     0.463375     0.335969  
std       0.142835     0.118736     0.127589     0.101135     0.029208  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%       0.415921     0.089744     0.102041     0.399420     0.328767  
50%       0.513417     0.153846     0.173469     0.462282     0.333333  
75%       0.610465     0.243590     0.265306     0.525145     0.342466  
max       1.000000     1.000000     1.000000     1.000000     1.000000

**[2.4]** Task: Create Train and Validate Datasets

In [26]:
# Drop Target and ID for X, and create target from y
df_train_id = df_train_cleaned['Id']
X = df_train_cleaned.drop(['Id', 'TARGET_5Yrs'], axis=1)
y = df_train_cleaned['TARGET_5Yrs']

In [27]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5256 entries, 0 to 5255
Data columns (total 19 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   GP       5256 non-null   float64
 1   MIN      5256 non-null   float64
 2   PTS      5256 non-null   float64
 3   FGM      5256 non-null   float64
 4   FGA      5256 non-null   float64
 5   FG%      5256 non-null   float64
 6   OREB     5256 non-null   float64
 7   DREB     5256 non-null   float64
 8   REB      5256 non-null   float64
 9   AST      5256 non-null   float64
 10  STL      5256 non-null   float64
 11  TOV      5256 non-null   float64
 12  3P Made  5256 non-null   float64
 13  3PA      5256 non-null   float64
 14  3P%      5256 non-null   float64
 15  FTM      5256 non-null   float64
 16  FTA      5256 non-null   float64
 17  FT%      5256 non-null   float64
 18  BLK      5256 non-null   float64
dtypes: float64(19)
memory usage: 780.3 KB


In [31]:
from sklearn.model_selection import train_test_split

In [30]:
# split the data in training and remaining dataset. 80:10:10 for train:valid:test
X_train, X_validate, y_train, y_validate  = train_test_split(X, y, train_size=0.8, shuffle = True, stratify=y, random_state = 8)

**[2.5]** Task: Check train and validate datasets have been distributed well

In [32]:
y.value_counts(normalize=True)

1    0.848364
0    0.151636
Name: TARGET_5Yrs, dtype: float64

In [33]:
y_train.value_counts(normalize=True)

1    0.848478
0    0.151522
Name: TARGET_5Yrs, dtype: float64

In [34]:
y_validate.value_counts(normalize=True)

1    0.847909
0    0.152091
Name: TARGET_5Yrs, dtype: float64

## 3. Create Baseline Model

**[3.1]** Task: Find the mode of the target variable and print it's value

In [35]:
y_mode = y_train.mode()
y_mode

0    1
dtype: int64

**[3.2]** Task: Create a numpy array called y_base filled with this value

In [36]:
y_shape = (len(y_train), 1)
y_base = np.full(y_shape, y_mode)

**[3.3]** Task: Import the accuracy score from sklearn

In [37]:
from sklearn.metrics import accuracy_score

**[3.4]** Task: Display the accuracy score of this baseline model

In [38]:
accuracy_score(y_train, y_base)

0.8484776403425309

The null accuracy is 83.35%
If we predict 1, we'll be correct 83.35% of the time

## 4. Random Forest - Hyperparameter Tuning with Grid Search and KFold

Train Random Forest Algorithm

1.   Use Grid Search to tune Hyper Parameters
2.   Use KFold with 5 Splits
3.   Fit Algorithm

**[4.1]** Task: Import GridSearchCV, KFold from sklearn.model_selection

In [39]:
from sklearn.model_selection import RandomizedSearchCV,GridSearchCV,KFold

**[4.2]** Create a dictionary containing the Randomized Grid Search Parameters

In [40]:
# Number of trees in random forest
n_estimators = np.linspace(50, 500, int((2000-100)/400) + 1, dtype=int)
# Number of features to consider at every split
max_features = ['auto', 'sqrt']
# Maximum number of levels in tree
max_depth = [1, 5, 10, 20, 50, 100]
# Minimum number of samples required to split a node
# min_samples_split = [int(x) for x in np.linspace(start = 2, stop = 10, num = 9)]
min_samples_split = [1, 2, 5, 10, 15, 20]
# Minimum number of samples required at each leaf node
min_samples_leaf = [2, 5, 8]
# Method of selecting samples for training each tree
bootstrap = [True, False]
# Criterion
criterion=['gini', 'entropy']
random_grid = {'n_estimators': n_estimators,
#                'max_features': max_features,
               'max_depth': max_depth,
               'min_samples_split': min_samples_split,
               'min_samples_leaf': min_samples_leaf,
               'bootstrap': bootstrap,
               'criterion': criterion}

In [41]:
random_grid

{'n_estimators': array([ 50, 162, 275, 387, 500]),
 'max_depth': [1, 5, 10, 20, 50, 100],
 'min_samples_split': [1, 2, 5, 10, 15, 20],
 'min_samples_leaf': [2, 5, 8],
 'bootstrap': [True, False],
 'criterion': ['gini', 'entropy']}

**[4.3]** Task: Create RF Model, and use Randomized Search to find GridSearch range

In [42]:
 from sklearn.ensemble import RandomForestClassifier

In [43]:
rf_base = RandomForestClassifier()
rf_random = RandomizedSearchCV(estimator = rf_base,
                               param_distributions = random_grid,
                               n_iter = 30, 
                               cv = 5,
                               verbose=2,
                               random_state=8, 
                               n_jobs = 4)

**[4.4]** Train Model

In [44]:
## Import up sound alert dependencies
from IPython.display import Audio, display

In [45]:
# Define a function to beep when done.
def allDone():
    display(Audio(url='https://upload.wikimedia.org/wikipedia/commons/0/05/Beep-09.ogg', autoplay=True))
    return

In [130]:
rf_random.fit(X_train, y_train)

# Beep when completed
allDone()

print(rf_random.best_params_)

Fitting 5 folds for each of 30 candidates, totalling 150 fits


/opt/conda/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:372: FitFailedWarning: 
25 fits failed out of a total of 150.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
10 fits failed with the following error:
Traceback (most recent call last):
  File "/opt/conda/lib/python3.9/site-packages/sklearn/model_selection/_validation.py", line 680, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/opt/conda/lib/python3.9/site-packages/sklearn/ensemble/_forest.py", line 450, in fit
    trees = Parallel(
  File "/opt/conda/lib/python3.9/site-packages/joblib/parallel.py", line 1043, in __call__
    if self.dispatch_one_batch(iterator):
  File "/opt/conda/lib/python3.9/site-packages/joblib/parallel.py", 

{'n_estimators': 50, 'min_samples_split': 5, 'min_samples_leaf': 8, 'max_depth': 20, 'criterion': 'entropy', 'bootstrap': True}


*print(rf_random.best_params_)*

{'n_estimators': 50, 'min_samples_split': 5, 'min_samples_leaf': 8, 'max_depth': 20, 'criterion': 'entropy', 'bootstrap': True}

In [46]:
rf_best = RandomForestClassifier(random_state=8, n_estimators=50, max_depth=5, min_samples_leaf=8, max_features=None).fit(X_train, y_train)

In [47]:
print("--- Accuracy Scores Best Fit Model ---")
print("Baseline: ",accuracy_score(y_train, y_base))
print("Train:    ",rf_best.score(X_train, y_train))
print("Validate: ",rf_best.score(X_validate, y_validate))

--- Accuracy Scores Best Fit Model ---
Baseline:  0.8484776403425309
Train:     0.8510941960038059
Validate:  0.8460076045627376


**Observation** Training Model is overfitting

**[4.5]** Tune Hyperparameters to better reduce overfitting.

1. Reduce Depth to and 
2. increase min samples per leaf to 10

In [48]:
rf_best_1 = RandomForestClassifier(random_state=8, n_estimators=100, max_depth=10, min_samples_leaf=10, max_features=None, class_weight=None).fit(X_train, y_train)

In [49]:
print("--- Accuracy Scores Best Fit Model ---")
print("Baseline: ",accuracy_score(y_train, y_base))
print("Train:    ",rf_best_1.score(X_train, y_train))
print("Validate: ",rf_best_1.score(X_validate, y_validate))

--- Accuracy Scores Best Fit Model ---
Baseline:  0.8484776403425309
Train:     0.8591817316841104
Validate:  0.8431558935361216


**Observations**
* Validation Dataset is not performing well for the model, indicating Model is overfitting on Train Dataset

**[4.6]** Task: Plot Confusion Matrix

Identify how well the model is performing at predicting the outcome

In [50]:
from sklearn.metrics import ConfusionMatrixDisplay
import matplotlib.pyplot as plt

In [51]:
plot_confusion_matrix(rf_best_1, X_train, y_train, cmap=plt.cm.Blues, normalize='true')
plot_confusion_matrix(rf_best_1, X_validate, y_validate, cmap=plt.cm.Blues, normalize='true')

NameError: name 'plot_confusion_matrix' is not defined

**Observation** Both Models is predicting alot of False Positives (FP) in both training and validation datasets

**[4.7]** Task: Calculate AUROC

In [52]:
from sklearn.metrics import roc_auc_score

In [53]:
rf_best_1_X_train_proba = rf_best_1.predict_proba(X_train)[:, 1]
rf_best_1_X_validate_proba = rf_best_1.predict_proba(X_validate)[:, 1]
print("rf_best_1_proba:", rf_best_1_X_train_proba)

rf_best_1_proba: [0.88523939 0.92917262 0.85111116 ... 0.4909883  0.64905682 0.79620799]


In [54]:
roc_auc_score(y_train, rf_best_1_X_train_proba)

0.9423293675366244

In [55]:
roc_auc_score(y_validate, rf_best_1_X_validate_proba)

0.6664027466367712

**Observations** 
* Training Dataset has a good AUROC prediction.
* Validation Dataset has a poor AUROC prediction. This indicates the model is over-fitting and not working well on unseen data
* The finaly Kaggle result AUROC 0.54153 continues this trend
* More work to be done

## 5. Random Forest - Feature Importance

**[5.1]** Task: Print feature importance

In [56]:
rf_best_1.feature_importances_

array([0.18937493, 0.06429554, 0.04712656, 0.03139556, 0.04351818,
       0.13156102, 0.0273911 , 0.04475946, 0.03914405, 0.04490803,
       0.02868494, 0.0343972 , 0.01590493, 0.05582963, 0.05546841,
       0.0280457 , 0.02815844, 0.06360572, 0.02643059])

**[5.2]** Task: Create a Data Frame with feature importance values

In [57]:
feat_imp_rf_df = pd.DataFrame(
    {'feature': X_train.columns,
     'feature_importance': rf_best_1.feature_importances_
     }
     )

**[5.3]** Task: Sort dataframe and reset index

In [58]:
feat_imp_rf_df = feat_imp_rf_df.sort_values('feature_importance', ascending=False).reset_index(drop=True)

**[5.4]** Task: Display the first 10 rows of important features

In [59]:
feat_imp_rf_df.head(10)

feature  feature_importance
0      GP            0.189375
1     FG%            0.131561
2     MIN            0.064296
3     FT%            0.063606
4     3PA            0.055830
5     3P%            0.055468
6     PTS            0.047127
7     AST            0.044908
8    DREB            0.044759
9     FGA            0.043518

**[5.5]**  Task: Create graph of Feature Importance

In [61]:
import altair as alt

ModuleNotFoundError: No module named 'altair'

In [152]:
alt.Chart(feat_imp_rf_df).mark_bar().encode(
    x='feature_importance',
    y=alt.Y("feature", sort='-x'),
    tooltip=['feature','feature_importance'])

alt.Chart(...)

## 6. Random Forest - Kaggle Competition Preparation

**[6.1]** Prepare Kaggle Input File using rf_best Random Forest Model

In [62]:
# Drop Target and ID for X, and create target from y
df_test_id = df_test_cleaned['Id']
X_test = df_test_cleaned.drop(['Id'], axis=1)

**[6.2]** Check Test Dataset is ready for predicting

In [63]:
X_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3799 entries, 0 to 3798
Data columns (total 19 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   GP       3799 non-null   float64
 1   MIN      3799 non-null   float64
 2   PTS      3799 non-null   float64
 3   FGM      3799 non-null   float64
 4   FGA      3799 non-null   float64
 5   FG%      3799 non-null   float64
 6   OREB     3799 non-null   float64
 7   DREB     3799 non-null   float64
 8   REB      3799 non-null   float64
 9   AST      3799 non-null   float64
 10  STL      3799 non-null   float64
 11  TOV      3799 non-null   float64
 12  3P Made  3799 non-null   float64
 13  3PA      3799 non-null   float64
 14  3P%      3799 non-null   float64
 15  FTM      3799 non-null   float64
 16  FTA      3799 non-null   float64
 17  FT%      3799 non-null   float64
 18  BLK      3799 non-null   float64
dtypes: float64(19)
memory usage: 564.0 KB


**[6.3]** Create y Prediction using the Test Dataset

In [72]:
y_predict = rf_best_1.predict_proba(X_test)

In [76]:
print(y_predict)

[[0.26613207 0.73386793]
 [0.30919053 0.69080947]
 [0.11440978 0.88559022]
 ...
 [0.22347663 0.77652337]
 [0.06077046 0.93922954]
 [0.25728307 0.74271693]]


In [77]:
print(y_predict[:,1])

[0.73386793 0.69080947 0.88559022 ... 0.77652337 0.93922954 0.74271693]


**[6.4]** Convert the Numpy Array into a Pandas Dataframe with a Column Name TARGET_5Yrs

In [78]:
# Convert numpy arrany y_predict values back to df_train_cleaned_scaled data frame 
y_predict = pd.DataFrame(y_predict[:,1], columns = ['TARGET_5Yrs'])

**[6.5]** Concatenate Player ID and Prediction into on Data Frame and Sort by Id

In [79]:
y_predict = pd.concat([df_test_id, y_predict], axis=1)

In [80]:
y_predict = y_predict.sort_values(by=['Id'])

In [82]:
print(y_predict)

         Id  TARGET_5Yrs
0         1     0.733868
2         3     0.885590
5         6     0.663792
8        14     0.771546
10       17     0.787609
...     ...          ...
3057  13770     0.881461
3058  13771     0.741607
3059  13774     0.696779
3063  13787     0.724787
3064  13792     0.828467

[3799 rows x 2 columns]


**[6.6]** Export Dataframe to CSV and save model

In [83]:
#joblib.dump(rf_best_1, "../models/nfragar_20220212_random_forest_v1_1.joblib", compress=3)
y_predict.to_csv("../data/external/submission_nfragar_20220212_random_forest_v1_2.csv", index = False)
y_predict.shape

(3799, 2)